app/src/main/java/eu/kanade/tachiyomi/data/track/kavita/KavitaApi.kt

```diff
@@ -0,0 +1,191 @@
+        for (sourceId in 1..3) {
+            val sourceSuffixID by lazy {
+                val key = "${"kavita_$sourceId"}/all/1" // Hardcoded versionID to 1
+                val bytes = MessageDigest.getInstance("MD5").digest(key.toByteArray())
+                (0..7).map { bytes[it].toLong() and 0xff shl 8 * (7 - it) }
+                    .reduce(Long::or) and Long.MAX_VALUE
+            }
+            val preferences: SharedPreferences by lazy {
+                Injekt.get<Application>().getSharedPreferences("source_$sourceSuffixID", 0x0000)
+            }
+            val prefApiUrl = preferences.getString("APIURL", "")!!
+
+            if (prefApiUrl.isNotEmpty()) {
+                if (prefApiUrl == getCleanedApiUrl(apiUrl)) {
+                    prefApiKey = Injekt.get<Application>()
+                        .getSharedPreferences("source_$sourceSuffixID", 0x0000)
+                        .getString("APIKEY", "")!!
+                    break
+                }
+            }
+        }
+        return prefApiKey
+    }
+
+    var apiUrl = ""
+    private fun getCleanedApiUrl(url: String): String {
+        /*Returns "<kavita_IP-Domain>/api/" from given url,*/
+        apiUrl = "${url.split("/api/").first()}/api"
+        return apiUrl
+    }
+    private fun getToken(url: String) {
+        /*
+         * Uses url to compare against each source APIURL's to get the correct custom source preference.
+         * Now having source preference we can do getString("APIKEY")
+         * Authenticates to get the token
+         * Saves the token in the var jwtToken
+         */
+        val cleanedApiUrl = apiUrl
+        val apiKey = getKavitaPreferencesApiKey(url)
+        if (apiKey.isEmpty()) {
+            logcat(LogPriority.WARN) { "Could not get api key" }
+            throw Exception("Could not load Api key")
+        }
+        val request = POST(
+            "$cleanedApiUrl/Plugin/authenticate?apiKey=$apiKey&pluginName=Tachiyomi-Kavita",
+            headersBuilder().build(),
+            "{}".toRequestBody("application/json; charset=utf-8".toMediaTypeOrNull()),
+        )
+        client.newCall(request).execute().use {
+            if (it.code == 200) {
+                jwtToken = it.parseAs<AuthenticationDto>().token
+            }
+            if (it.code == 401) {
+                logcat(LogPriority.WARN) { "Unauthorized / api key not valid:\nCleaned api URL:${cleanedApiUrl}\nApi key is empty:${apiKey.isEmpty()}\n" }
+                throw Exception("Unauthorized / api key not valid")
+            }
+            if (it.code == 500) {
+                logcat(LogPriority.WARN) { "Error fetching jwd token:\nCleaned api URL:${cleanedApiUrl}\nApi key is empty:${apiKey.isEmpty()}\n" }
+                throw Exception("Error fetching jwd token")
+            }
+        }
+    }
+    private fun getApiVolumesUrl(url: String): String {
+        return "${url.split("/api/").first()}/api/Series/volumes?seriesId=${getIdFromUrl(url)}"
```

Comment:
Why not use `apiUrl` or `getCleanedApiUrl`?
```suggestion
        return "${apiUrl}/Series/volumes?seriesId=${getIdFromUrl(url)}"
```

```suggestion
        return "${getCleanedApiUrl(url)}/Series/volumes?seriesId=${getIdFromUrl(url)}"
```